core/server/worker/src/main/java/alluxio/worker/block/annotator/DefaultBlockIterator.java

```diff
@@ -0,0 +1,420 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.block.annotator;
+
+import alluxio.collections.ConcurrentHashSet;
+import alluxio.collections.Pair;
+import alluxio.worker.block.AbstractBlockStoreEventListener;
+import alluxio.worker.block.BlockMetadataManager;
+import alluxio.worker.block.BlockStoreEventListener;
+import alluxio.worker.block.BlockStoreLocation;
+import alluxio.worker.block.meta.StorageDir;
+import alluxio.worker.block.meta.StorageTier;
+
+import com.google.common.collect.Iterators;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * The default {@link BlockIterator} implementation that integrates with the
+ * {@link BlockMetadataManager}.
+ */
+public class DefaultBlockIterator implements BlockIterator {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultBlockIterator.class);
+
+  /** Map of sorted block set collections per directory. */
+  private final HashMap<BlockStoreLocation, SortedBlockSet<BlockSortedField>> mPerDirOrderedSets;
```

Comment:
It seems possible this will be modified concurrently during iteration?

Reply:
Correct.

Reply:
I think you mean the `storageLost()` callback. Nice catch.
Updated to a concurrent-map.